test/e2e_node/pod_status_test.go

```diff
@@ -0,0 +1,150 @@
+				Name:   "initialized-pod",
+				Labels: podLabels,
+			},
+			Spec: v1.PodSpec{
+				RestartPolicy: v1.RestartPolicyAlways,
+				InitContainers: []v1.Container{
+					{
+						Name:  init,
+						Image: busyboxImage,
+						Command: ExecCommand(init, execCommand{
+							Delay:    30,
+							ExitCode: 0,
+						}),
+					},
+					{
+						Name:  restartableInit,
+						Image: busyboxImage,
+						Command: ExecCommand(restartableInit, execCommand{
+							Delay:    300,
+							ExitCode: 0,
+						}),
+						RestartPolicy: &containerRestartPolicyAlways,
+					},
+				},
+				Containers: []v1.Container{
+					{
+						Name:  regular,
+						Image: busyboxImage,
+						Command: ExecCommand(regular, execCommand{
+							Delay:    300,
+							ExitCode: 0,
+						}),
+					},
+				},
+			},
+		}
+		preparePod(pod)
+
+		client := e2epod.NewPodClient(f)
+		pod = client.Create(ctx, pod)
+
+		ginkgo.By("Waiting for the regular init container to be started")
+		pod, err := waitForPodInitContainerStarted(ctx, f.ClientSet, pod.Namespace, pod.Name, init, f.Timeouts.PodStart)
+		framework.ExpectNoError(err)
+		if pod.Status.Phase != v1.PodPending {
+			framework.Failf("pod should be pending during the execution of the init container")
+		}
+
+		ginkgo.By("Getting the current pod sandbox ID")
+		rs, _, err := getCRIClient()
+		framework.ExpectNoError(err)
+
+		sandboxes, err := rs.ListPodSandbox(ctx, &runtimeapi.PodSandboxFilter{
+			LabelSelector: podLabels,
+		})
+		framework.ExpectNoError(err)
+		gomega.Expect(sandboxes).To(gomega.HaveLen(1))
+		podSandboxID := sandboxes[0].Id
+
+		ginkgo.By("Stopping the kubelet")
+		restartKubelet := stopKubelet()
+		gomega.Eventually(ctx, func() bool {
+			return kubeletHealthCheck(kubeletHealthCheckURL)
+		}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeFalseBecause("kubelet should be stopped"))
+
+		ginkgo.By("Stopping the pod sandbox to simulate the node reboot")
+		err = rs.StopPodSandbox(ctx, podSandboxID)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Restarting the kubelet")
+		restartKubelet()
+		gomega.Eventually(ctx, func() bool {
+			return kubeletHealthCheck(kubeletHealthCheckURL)
+		}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeTrueBecause("kubelet should be started"))
+
+		ginkgo.By("Waiting for the regular init container to be started after the node reboot")
+		pod, err = waitForPodInitContainerStarted(ctx, f.ClientSet, pod.Namespace, pod.Name, init, f.Timeouts.PodStart)
+		framework.ExpectNoError(err)
+		if pod.Status.Phase != v1.PodPending {
+			framework.Failf("pod should be pending during the execution of the init container after the node reboot")
+		}
+	})
+})
+
+// waitForPodInitContainerStarted waits for the given Pod init container to start, after a successful run of the startupProbe.
+func waitForPodInitContainerStarted(ctx context.Context, c clientset.Interface, namespace, podName, initContainer string, timeout time.Duration) (*v1.Pod, error) {
```

Comment:
I think it's a bit strange that the `waitForPodInitContainerStarted()` function returns the Pod spec, given the name of the function. It might make more sense to add an Init container version of `e2epod.WaitForPodContainerStarted()` and then fetch the Pod spec with `client.Get()` afterward.

Reply:
Thank you very much for your careful review. There is indeed an issue with waitForPodInitContainerStarted.

Reply:
The reason the test may not have taken effect could be due to the following code, which should wait for the regular container to be started. In any case, I accept your suggestion, thank you very much.
> 		ginkgo.By("Waiting for the regular init container to be started")
>		pod, err := waitForPodInitContainerStarted(ctx, f.ClientSet, pod.Namespace, pod.Name, init, f.Timeouts.PodStart)

Reply:
> I think it's a bit strange that the `waitForPodInitContainerStarted()` function returns the Pod spec, given the name of the function. It might make more sense to add an Init container version of `e2epod.WaitForPodContainerStarted()` and then fetch the Pod spec with `client.Get()` afterward.

donw